src/Deferred.php

```diff
@@ -5,58 +5,22 @@
 namespace GraphQL;
 
 use GraphQL\Executor\Promise\Adapter\SyncPromise;
-use SplQueue;
-use Throwable;
 
-class Deferred
+class Deferred extends SyncPromise
 {
-    /** @var SplQueue|null */
-    private static $queue;
-
-    /** @var callable */
-    private $callback;
-
-    /** @var SyncPromise */
-    public $promise;
-
-    public function __construct(callable $callback)
-    {
-        $this->callback = $callback;
-        $this->promise  = new SyncPromise();
-        self::getQueue()->enqueue($this);
-    }
-
-    public static function getQueue() : SplQueue
-    {
-        if (self::$queue === null) {
-            self::$queue = new SplQueue();
-        }
-
-        return self::$queue;
-    }
-
-    public static function runQueue() : void
-    {
-        $queue = self::getQueue();
-        while (! $queue->isEmpty()) {
-            /** @var self $dequeuedNodeValue */
-            $dequeuedNodeValue = $queue->dequeue();
-            $dequeuedNodeValue->run();
-        }
-    }
-
-    public function then($onFulfilled = null, $onRejected = null)
+    /**
+     * @param callable() : mixed $executor
+     */
+    public static function create(callable $executor) : self
```

Comment:
Are we fine with the fact that child class cannot override it to create self instance? This won't be possible:

```php

class A extends Deferred
{
    public static function create(callable $executor) : self
    {
       // something
        return parent::create($executor);
    }
}
```

I'm fan of final classes, just asking.

Reply:
It is not meant to be extended. But if you really want to, you can still override the constructor (you shouldn't, really).